pkg/chartutil/save.go

```diff
@@ -231,10 +237,9 @@ func writeTarContents(out *tar.Writer, c *chart.Chart, prefix string) error {
 func writeToTar(out *tar.Writer, name string, body []byte) error {
 	// TODO: Do we need to create dummy parent directory names if none exist?
 	h := &tar.Header{
-		Name:    filepath.ToSlash(name),
-		Mode:    0644,
-		Size:    int64(len(body)),
-		ModTime: time.Now(),
```

Comment:
Removing `ModTime` will reintroduce behavior described in #4158, which was fixed in #4161 and ported to V3 in #7272

Reply:
IMHO, the current implementation of modTime is also wrong because it just sets the time when helm package was run. If it wants to be correct, it should set the original template files timestamp.

Also, zero-ing out timestamps is a common pattern for reproducible builds https://github.com/GoogleContainerTools/distroless/issues/242

Reply:
Regardless, @baijum is right. We tried this, and someone reported it as a bug that was addressed with #4161. To go back and re-implement this again would be repeating past mistakes.

Reply:
May be this can be added as a flag to helm package command for those of us want to have reproducible charts.

Reply:
https://github.com/helm/helm/issues/3612#issuecomment-525340295 suggests the issue is not due to timestamps, but instead the sorted order of the files within the tar file.

Can you please look further into the suggestions made in that thread before coming up with a solution. Even if you strip mtime fields from the archive, you may still not produce a reproducible build. We've looked into this issue multiple times and discovered it's not the mtime of the original files causing the issue here. Or at least, there are multiple layers causing this issue, and to solve it we need to address them all.

We tested this theory in another area and discovered stripping mtime from the archive does not resolve the issue. See https://github.com/helm/helm/pull/8216#pullrequestreview-437935656.

Reply:
Can we add a backwards-compatible flag, like `helm package --timestamp` where `timestamp` defaults to `true`, but users can set it to false for reproducibility on that level.

Reply:
The first step would be to write a test to identify the symptom. Then write a fix to alleviate the symptom.

It is possible that [we only need to sort the files in the archive](https://github.com/helm/helm/pull/8216#pullrequestreview-437935656) to fix the issue, which means there would not be any need for stripping file mtimes (and therefore no need for a `--timestamp` flag).

We did test the theory of stripping file mtimes before and [that did not allay the symptom](https://github.com/helm/helm/pull/8216#pullrequestreview-437935656). @jdolitsky can also attest to this as he and his coworker tried to fix this for `helm chart save`. So I'm not confident that stripping file mtimes will fix the problem.


Let's get some tests in first before we start providing solutions.

Reply:
This should be tried without the timestamp going to 0. But, I suspect any timestamp difference will cause a different hash. A different time stored in the tar when everything else is the same is still different content that would produce a different hash.

In another part of the PR there is sorting on the content. The order of the files matters and I hope the change here solves the issue. I have not checked, though.

If the timestamp needs to be 0 then that would need to be hidden behind a flag. We can't break backwards compatibility. I would suggest a name like `--reproducible-build`.

Reply:
Please be aware that [Git doesn't preserve modification time](https://git.wiki.kernel.org/index.php/Git_FAQ#Why_isn.27t_Git_preserving_modification_time_on_files.3F) on files. So an Application Distributor cannot create a reproducible build by preserving the original file's modification time because the modified time will be different whenever the chart source checks out. I think using a flag (`--reproducible-build`) and setting timestamp to 0 looks acceptable to me—some tests to avoid regression would be great.